Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialVersionUpdate.cpp

```diff
@@ -208,44 +249,67 @@ namespace AZ
             return renamed;
         }
 
-        bool MaterialVersionUpdate::CastToExpectedValue(
-            const Name& propertyId, MaterialPropertyValue& value, const MaterialPropertiesLayout* materialPropertiesLayout,
-            AZStd::function<void(const char*)> onError)
+        MaterialVersionUpdate::PropertyHelper::PropertyHelper(
+            AZStd::function<bool(AZ::Name&)> applyAllPropertyRenames,
+            const MaterialPropertiesLayout* materialPropertiesLayout)
+            : m_applyAllPropertyRenames(applyAllPropertyRenames), m_materialPropertiesLayout(materialPropertiesLayout)
         {
+        }
+
+        bool MaterialVersionUpdate::PropertyHelper::CastToExpectedType(
+            const Name& providedPropertyId, MaterialPropertyValue& value, AZStd::function<void(const char*)> onError) const
+        {
+            AZ_Assert(m_materialPropertiesLayout != nullptr, "PropertyHelper is not properly initialized");
+
+            // Update property id to latest name
+            Name propertyId(providedPropertyId);
+            ApplyAllPropertyRenames(propertyId);
```

Comment:
I bet we could pass in a material type version number to CastToExpectedType and ApplyAllPropertyRenames. Then it would only apply renames for the versions up to that point. This would be a bit more predictable for the users, knowing it won't apply to future renames. What do you think?

(I don't consider this a blocking issue, just nice to have)

Reply:
I feel like this might complicate the process for not much gain. The goal is mostly to reach the final name as quickly as possible, because only then can there be a lookup in the material properties layout.

In particular, errors in material names or incompatible types in setValue actions would otherwise only be spotted after the very last update (because only then the names would map correctly to the material properties layout), which would lead to less-informative error messages (or tagging along extra information throughout the process) instead of already catching and reporting such errors directly when processing the action that causes it.
The update currently already reports the original name that was used in such invalid setValue update actions, as well as the final name that it eventually maps to if this is different. An added bonus is that the setValue rules are thus robust to using names that are not exactly 'right' for that particular update's 'toVersion' -- as long as it was *any* name that existed for that property, it will work. Nonetheless, a point could be made that in such 'mismatched' case it could be useful to trigger a warning just in case something was unintentional, which would indeed require something like providing the 'current' version and probably some extra checking to only report this once per property.

Reply:
Yeah, you're right, I was on the wrong train of thought here.